docs/integrations/engines/clickhouse.md

```diff
@@ -0,0 +1,295 @@
+
+``` sql linenums="1" hl_lines="4-6"
+MODEL (
+    name my_schema.my_log_table,
+    kind full,
+    physical_properties (
+        index_granularity = 128
+    )
+);
+
+select
+    *
+from other_schema.other_table;
+```
+
+Note that there is an `=` between the `index_granularity` key name and value `128`.
+
+### Query settings
+
+Clickhouse settings may be specified directly in a model's query with the `SETTINGS` keyword.
+
+This example demonstrates setting the `join_use_nulls` setting to `1`:
+
+``` sql linenums="1" hl_lines="9"
+MODEL (
+    name my_schema.my_log_table,
+    kind full,
+);
+
+select
+    *
+from other_schema.other_table
+SETTINGS join_use_nulls = 1;
+```
+
+Multiple settings may be specified in a query with repeated use of the `SETTINGS` keyword: `SELECT * FROM other_table SETTINGS first_setting = 1 SETTINGS second_setting = 2;`.
+
+#### Usage by SQLMesh
+
+The Clickhouse setting `join_use_nulls` affects the behavior of SQLMesh SCD models and table diffs. This section describes how SQLMesh uses query settings to control that behavior.
+
+^^Background^^
+
+In general, table `JOIN`s can return empty cells for rows not present in both tables.
+
+For example, consider `LEFT JOIN`ing two tables `left` and `right`, where the column `right_column` is only present in the `right` table. Any rows only present in the `left` table will have no value for `right_column` in the joined table.
+
+In other SQL engines, those empty cells are filled with `NULL`s.
+
+In contrast, Clickhouse fills the empty cells with data type-specific default values (e.g., 0 for integer column types). It will fill the cells with `NULL`s instead if you set `join_use_nulls` to `1`.
+
+^^SQLMesh^^
+
+SQLMesh automatically generates SQL queries for both SCD Type 2 models and table diff comparisons. These queries include table `JOIN`s and calculations based on the presence of `NULL` values.
+
+Because those queries expect `NULL` values in empty cells, SQLMesh automatically adds `SETTINGS join_use_nulls = 1` to the generated SCD and table diff SQL code.
+
+The SCD model definition query is embedded as a CTE in the full SQLMesh-generated query. If run alone, the model definition query would use the Clickhouse server's current `join_use_nulls` value.
+
+If that value is not `1`, the SQLMesh setting on the outer query would override the server value and produce incorrect results.
+
+Therefore, SQLMesh uses the following procedure to ensure the model definition query runs with the correct `join_use_nulls` value:
+
+- If the model query sets `join_use_nulls` itself, do nothing
+- If the model query does not set `join_use_nulls` and the current server `join_use_nulls` value is `1`, do nothing
+- If the model query does not set `join_use_nulls` and the current server `join_use_nulls` value is `0`, add `SETTINGS join_use_nulls = 0` to the CTE model query
+    - All other CTEs and the outer query will still execute with a `join_use_nulls` value of `1`
```

Comment:
Are you asking about NULL not being used for missing rows in the user query only, or hypothetically about what would happen if we didn't inject the setting?

Assuming the former, they should work - every comparison they make is one of:
- Are these two non-NULL values equal?
- Is one of these values NULL and the other isn't?
- Are these both NULL?

If the user query does not use NULL for missing rows, all the missing rows their query returns will have values, so our comparisons are of two non-NULL values.

We could have an issue if the user query had a join, and one of the non-key columns in their join was used in the SCD primary key:
  - If there was no match for a row in their query, that column would get a default value instead of NULL.
  - SCD would see that default value as a "real" key
  - If the row subsequently had a match in their query, the key would modify
  - SCD would see that as a new key and treat it as a new row, rather than as a row update

I can add that to the docs.

Can you think of any other problematic scenarios?

Reply:
I didn't have an exact example in mind, I think I was under the assumption that SCD2 required `NULL`s instead of values for missing rows. So then I thought "if the user's _not_ using `NULL`s for missing rows, will the SCD2 logic work correctly? Given that we won't set `join_use_nulls` ourselves in this case to respect the model query's semantics. Does this make sense?